api: Add hcpOpenShiftVersions proxy resource type

mbarnes opened the pull request:
[ARO-17700 - REST Specs API: Support Proxy ResourceType For OCP Versions](https://issues.redhat.com/browse/ARO-17700)

### What

Adds a new proxy resource type named `hcpOpenShiftVersions` that represents available HCP OpenShift versions ~~and upgrade paths~~.

Here's an illustration of a resource JSON payload:
```
{
    "properties": {
        "channelGroup": "stable",
        "enabled": true,
        "endOfLifeTimestamp": "2032-12-22T10:00:00Z",
    }
}
```

### Why

This is a "top-level" resource type within the `Microsoft.RedHatOpenShift` namespace so we have to be mindful of ARO Classic resource type names.

Whereas ARO Classic uses `openShiftClusters` and `openShiftVersions`, `hcpOpenShiftVersions` seemed like the natural choice of name to pair with `hcpOpenShiftClusters`.

### Special notes for your reviewer

I couldn't find any design docs about how this is meant to be exposed in ARO-HCP, so I'm working from a bunch of assumptions.  Some fields from the [source model](https://github.com/openshift-online/ocm-api-model/blob/1e9afd48845ab1105609e4438dae96f0afa5d3e6/clientapi/arohcp/v1alpha1/version_type.go#L38-L59) are obviously not applicable to ARO-HCP, others are less clear.

**Open questions:**

- Will `default` and `enabled` in the ARM API map to `default` and `enabled` or to `hosted_control_plane_default` and `hosted_control_plane_enabled` in the OCM model?

- `image_overrides` from the OCM model is not currently applicable to Azure.  I'm assuming this field is not needed for ARO-HCP unless I'm told otherwise.

- I also assume `wif_enabled` is not applicable to ARO-HCP unless I'm told otherwise.

_**Update:**_ Decision made to drop `releaseImage` and  `availableUpgrades` entirely so the questions below are moot.

~~- Release images and architectures is where I need the most clarification.~~

  1. ~~Do we need to be publishing release image references in the ARM API at all?~~
  
      ~~Cluster Service's `/api/aro_hcp/v1alpha1/versions` endpoint is returning release images in `quay.io` instead of `*.azurecr.io`.  That seems inappropriate for Azure.  Is that going to change?~~
  
  2. ~~If release images do need to be included, what architectures are we supporting?  I see `amd64`, `arm64` and `multi` as options.~~
  
      ~~I get the impression from the OCM model that the amd64-specific `release_image` field (and maybe `available_upgrades`) came before the `release_images` list, and that this is API tech debt.  In other words, ideally the OCM model would just be~~
      ```
      release_images:
          amd64:
              available_upgrades: [string]
              release_image: string
          arm64:
              available_upgrades: [string]
              release_image: string
          multi:
              available_upgrades: [string]
              release_image: string
      ```
      ~~with no separate `release_image` or `available_upgrades` fields?  If I'm right about all that, is that how we should structure the ARM API?~~

mbarnes commented:
(Reminder to myself to generate API examples once we get the API itself figured out.)

machi1990 commented:
> Do we need to be publishing release image references in the ARM API at all?

How does aro-classic handles this? 

> Cluster Service's /api/aro_hcp/v1alpha1/versions endpoint is returning release images in quay.io instead of *.azurecr.io. That seems inappropriate for Azure. Is that going to change?
 
 @tjungblu would you be able to chime in on this? 
 
@mbarnes  My personal opinion on this, the ACR isn't a public one and I don't know how useful it is to return the info to the end user if they can't even access the images themselves; so I'd say this info remain internal only and not exposed to the end user i.e not exposed via the ARM API.

>  If release images do need to be included, what architectures are we supporting? I see amd64, arm64 and multi as options.

We only support `multi` archi in CS for aro-hcp. So I'd have something like,
`architectures` or `supported_architectures` with `multi` underneath it.

```json
"supported_architectures": {
   "multi": {
     "available_upgrades": [....]
 }
}
```

That will keep the field extensible. 

> I also assume wif_enabled is not applicable to ARO-HCP unless I'm told otherwise.

This isn't applicable. 

> image_overrides from the OCM model is not currently applicable to Azure. I'm assuming this field is not needed for ARO-HCP unless I'm told otherwise.

No, this isn't applicable for now as it's only gcp/aws for now. But I'd assume (don't take my words in this) in the future we'll put the different VM images in this section.

> Will default and enabled in the ARM API map to default and enabled or to hosted_control_plane_default and hosted_control_plane_enabled in the OCM model?

ARM API `default` will map to `default`
ARM API `enabled` will map to `ocm.enabled && ocm.hosted_control_plane_enabled`
@tjungblu any thoughts on this?

machi1990 commented:
> > > Do we need to be publishing release image references in the ARM API at all?
> > 
> > 
> > How does aro-classic handles this?
> 
> Turns out ARO Classic [exposes the Quay.io image reference](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config?path=/deploy/prod-config.yaml&version=GBmaster&line=2875&lineEnd=2885&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) as well.
> 

Thanks for checking.

> @machi1990 Ok, so, CS returns a release image for both `amd64` and `multi`.
> 
> ```
> {
>   "kind":"Version",
>   "id":"openshift-v4.18.1",
>   "href":"/api/clusters_mgmt/v1alpha1/versions/openshift-v4.18.1",
>   "raw_id":"4.18.1",
>   "enabled":true,
>   "default":true,
>   "channel_group":"stable",
>   "rosa_enabled":true,
>   "hosted_control_plane_enabled":true,
>   "hosted_control_plane_default":false,
>   "gcp_marketplace_enabled":false,
>   "release_image":"quay.io/openshift-release-dev/ocp-release@sha256:d9c985464c0315160971b3e79f5fbec628d403a572f7a6d893c04627c066c0bb",
>   "release_images": {
>     "multi": {
>       "release_image":"quay.io/openshift-release-dev/ocp-release@sha256:a525c21dc0f92a65911bf073d1d284a988a866ed2b0952ffe63608b041504190"
>     }
>   },
>   "wif_enabled":false
> }
> ```
> 
> If I understand correctly, we'll disregard the `amd64` image and return a `supported_architectures` (name is TBD) section that looks like:
> 

Yes.  From the returned info from CS, RP can disregard `amd64` and have a section that only returns `multi` as supported architecture for now since that's what is currently handled. 
 
> ```
> "supported_architectures": [
>     {
>         "name": "multi",
>         "release_image": "quay.io/openshift-release-dev/ocp-release@sha256:a525c21dc0f92a65911bf073d1d284a988a866ed2b0952ffe63608b041504190",
>         "available_upgrades": [....]
>     }
> ]
> ```
> 
> (ARM doesn't like dictionaries, so I made it a list with `"name"` as the identifier field.)

👍 That looks good to me.  A structure like that will make it easier (as its flexible enough) to add new architectures in the ARM API if we are to support them in the future.

mbarnes commented:
> From the returned info from CS, RP can disregard amd64 and have a section that only returns multi as supported architecture for now since that's what is currently handled.

This got me thinking... if available upgrades can vary by architecture then I guess the RP should be using `.release_images["multi"].available_upgrades` instead of `.available_upgrades` in cluster and node pool version info?

miguelsorianod commented on api/redhatopenshift/HcpCluster.Management/hcpVersion.tsp:
```diff
@@ -0,0 +1,54 @@
+  channelGroup: string;
+
+  /** Default indicates if this version will be selected when a cluster is created without explicitly specifying the version. */
+  default: boolean;
```
cc @jharrington22

mbarnes replied:
Decision was reached in the API meeting to remove `"default"` for now.

FWIW, when the time comes to expose defaults, I'm partial to using a special resource name `"default"` that acts like a soft link to another resource instance, over a `"default"` boolean in every instance.  That way there's no possibility of multiple resources having `default: true`.  A GET to `../hcpOpenShiftVersions/default` would return the same as a GET to `../hcpOpenShiftVersions/4.20.1` if 4.20.1 were the default version.

mbarnes commented:
> In the cluster response, if .multi_arch_enabled==true then it is multi. Otherwise it is amd64. For now, all aro-hcp are multi.

This might be getting too much in the weeds, but what about `arm64`?  Hypothetically, if `arm64`, `amd64`, and `multi` were all supported options and `.multi_arch_enabled==false`, how would I distinguish between `arm64` and `amd64`?

mbarnes commented:
Updated commit removes the enabled flag and release image.  Still kept the `supported_architectures` list where `available_upgrades` resides.  Updated the PR description with the current model for easy review.  I'll work on generating API examples next.

jboutaud commented:
We will only support multi-arch in ARO HCP. Multi-arch means you can have nodes of various architecture in the cluster. ROSA didn't start multi-arch and thus has additional parameters that aren't needed. I don't think including architecture is relevant as a parameter here. 

Upgrade path is a bit more complex as yes, there a number of conditions that could impact upgrade path availability and some will be cluster specific. For the purpose of this API, let's keep it simple without taking into account the conditions where an upgrade wouldn't be possible.

machi1990 reviewed (commented):
lgtm
